streams.go

```diff
@@ -0,0 +1,311 @@
+package lksdk
+
+import (
+	"bytes"
+	"sync"
+	"time"
+
+	protocol "github.com/livekit/protocol/livekit"
+)
+
+type baseStreamInfo struct {
+	Id         string
+	MimeType   string
+	Topic      string
+	Timestamp  int64
+	Size       *uint64
+	Attributes map[string]string
+}
+
+type ByteStreamInfo struct {
+	*baseStreamInfo
+	Name *string
+}
+
+type TextStreamInfo struct {
+	*baseStreamInfo
+}
+
+const (
+	STREAM_CHUNK_SIZE = 15_000
+)
+
+type StreamTextOptions struct {
+	Topic                 string
+	DestinationIdentities []string
+	StreamId              *string
+	ReplyToStreamId       *string
+	TotalSize             *uint64
+	Attributes            map[string]string
+	OnProgress            *func(progress float64)
+}
+
+type StreamBytesOptions struct {
+	Topic                 string
+	MimeType              string
+	DestinationIdentities []string
+	StreamId              *string
+	TotalSize             *uint64
+	Attributes            map[string]string
+	OnProgress            *func(progress float64)
+	FileName              *string
+}
+
+type baseStreamWriter[T any] struct {
+	onWrite func(data T) (int, error)
+	onClose func() error
+}
+
+func newBaseStreamWriter[T any](onWrite func(data T) (int, error), onClose func() error) *baseStreamWriter[T] {
+	return &baseStreamWriter[T]{
+		onWrite: onWrite,
+		onClose: onClose,
+	}
+}
+
+// is it okay for TextStreamWriter to also return number of bytes written?
+func (w *baseStreamWriter[T]) Write(data T) (int, error) {
+	return w.onWrite(data)
+}
+
+func (w *baseStreamWriter[T]) Close() error {
+	return w.onClose()
+}
+
+func writeStreamBytes(data []byte, e *RTCEngine, streamId string, destinationIdentities []string, totalSize *uint64, onProgress *func(progress float64)) (int, error) {
+	chunkIndex := uint64(0)
+	bytesWritten := 0
+	for i := 0; i < len(data); i += STREAM_CHUNK_SIZE {
+		end := i + STREAM_CHUNK_SIZE
+		if end > len(data) {
+			end = len(data)
+		}
+
+		// this is a blocking call, but, if we call it in a goroutine internally,
+		// error propagation will require some blocking code which ruins the point
+		// is it best for the user to call this in a goroutine?
+		// or do we not block on buffer status low and just let the user handle it?
+		err := e.waitForBufferStatusLow(protocol.DataPacket_RELIABLE)
+		if err != nil {
+			return bytesWritten, err
+		}
+
+		if err := e.publishStreamChunk(&protocol.DataStream_Chunk{
+			StreamId:   streamId,
+			Content:    data[i:end],
+			ChunkIndex: chunkIndex,
+		}, destinationIdentities); err != nil {
+			return bytesWritten, err
+		}
+
+		if onProgress != nil && totalSize != nil {
+			progress := float64(i) / float64(*totalSize)
+			(*onProgress)(progress)
```

Comment:
Wondering if `onProgress` can be used to indicate error too. Is `onProgress` going back to the user?

Reply:
OnProgress is optional, user might not set it